Rewrite main server to be more conventional.

sleepdeprecation opened the pull request:
Instead of using just a ton of reflection and things otherwise
considered "not best practices", the main server now follows java
conventions a whole lot more.

What still needs to be done is a rewrite of most of the javadocs,
because basically all of the internals changed, and they're not up to
date.

This commit was maybe a bit too large...

sleepdeprecation committed:
Rewrite main server to be more conventional.

Instead of using just a ton of reflection and things otherwise
considered "not best practices", the main server now follows java
conventions a whole lot more.

What still needs to be done is a rewrite of most of the javadocs,
because basically all of the internals changed, and they're not up to
date.

This commit was maybe a bit too large...

sleepdeprecation commented:
This probably shouldn't be merged in until the documentation is more up to date, and the other changes being considered are added in.

This is mainly looking for comments on the code (both quality and functionality).

sleepdeprecation commented on src/main/java/httpserver/HTTPHandler.java:
```diff
@@ -22,232 +22,187 @@
+     *
+     * @see HTTPHandler#addGET
+     * @see HTTPHandler#addPOST
+     * @see HTTPHandler#addDELETE
```
`addGET`, `addPOST`, and `addDELETE` have been renamed `get`, `post`, and `delete`.

sleepdeprecation commented on src/main/java/httpserver/HTTPHandler.java:
```diff
@@ -22,232 +22,187 @@
+     *
+     * If there is not exact match, the `*` path is used. If you don't have a `*`
+     * catchall route, a 501 (Not implemented) is sent to the client.
+     *
```
Should something about not needing to override this, and that it's highly suggested you leave this as-is be added here?

sleepdeprecation commented on src/main/java/httpserver/HTTPHandler.java:
```diff
@@ -22,232 +22,187 @@
+    public void handle(HTTPRequest request, HTTPResponse response) {
+	String httpRequestType = request.getRequestType().toUpperCase();
+	if (!routes.containsKey(httpRequestType)) {
+	    response.message(404, "No " + httpRequestType + " routes exist.");
```
Should this send a `501` instead of a `404`? There technically is no method at that route.

sleepdeprecation commented on src/main/java/httpserver/HTTPHandler.java:
```diff
@@ -22,232 +22,187 @@
+     *
+     * Additionally, primitives are not permited, because they're not classes in
+     * the java.lang library. The three most common parameter types are String,
+     * Integer, and Double.
```
Almost all of this is inaccurate now. And the future documentation should probably be attached to the `addRoute` method, with the rest of them saying `@see HTTPHandler#addRoute`.

Paths should now come in `/path/to/thing/{variableName}` form, instead of `/path/to/thing/{variableType}`. There's also the option to accept `/path/to/thing/{*}`, which acts as a catchall list (and puts anything in `request.varargs`), which will match like the `{Class...}` matchers did before, but only if it's at the end of the path.

Also, it looks like we never included `{Class...}` documentation...

sleepdeprecation commented on src/main/java/httpserver/HTTPHandler.java:
```diff
@@ -22,232 +22,187 @@
+     * @see HTTPRequest
+     */
+    public void get(String path, Route route) throws HTTPException {
+		addRoute(HTTPRequest.GET_REQUEST_TYPE, path, route);
```
Side comment: should there be a `HTTPMethods` class, which is just a collection of constants.

Basically you'd say `HTTPMethods.GET` instead of `HTTPRequest.GET_REQUEST_TYPE`.

sleepdeprecation commented on src/main/java/httpserver/HTTPHandler.java:
```diff
@@ -22,232 +22,187 @@
  * @see MessageHandler
  */
 public abstract class HTTPHandler {
```
Should we rename our classes to follow standard convention? Meaning they'd all change from `HTTPClass` to `HttpClass`. This would be `HttpHandler`.

mpeterson2 replied:
:+1:

sleepdeprecation commented on src/main/java/httpserver/HTTPHandler.java:
```diff
@@ -22,232 +22,187 @@
+			return;
+		}
+
+		finalMethod.invoke(request, response, request.getSplitPath());
```
Indentation jankiness (from 83 to here).

sleepdeprecation commented on src/main/java/httpserver/HTTPHandler.java:
```diff
@@ -22,232 +22,187 @@
+     * @see HTTPHandler#addDELETE
+     */
+    public void post(String path, Route route) throws HTTPException {
+		addRoute(HTTPRequest.POST_REQUEST_TYPE, path, route);
```
More indentation jankiness. This should be looked at more carefully over the whole project.

dkuntz2 replied:
More indentation jankiness. This should be looked at more carefully over the whole project.

sleepdeprecation commented on src/main/java/httpserver/HTTPHandler.java:
```diff
@@ -22,232 +22,187 @@
+     * @param route	    The Route to be called at said path.
+     *
+     * @throws HTTPException  When you do bad things.
+     */
```
This documentation should be completely rewritten (see note on line 119).

dkuntz2 commented on src/main/java/httpserver/HTTPRequest.java:
```diff
@@ -25,509 +25,502 @@
+            Issue 12: https://github.com/dkuntz2/java-httpserver/issues/12
+            RFC 2616#4.2: http://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2
+            */
+        for (String line = input.readLine(); line != null && !line.isEmpty(); line = input.readLine()) {
```
Does `BufferedReader#readLine` return null, or just say it's empty? (Do we need the `line != null` check?)

sleepdeprecation commented on src/main/java/httpserver/HTTPRequest.java:
```diff
@@ -25,509 +25,502 @@
  * @see HTTPHandler
  */
 public class HTTPRequest implements Runnable {
```
We should consider getting rid of all of communication layer stuff, and just transition to having a simplifying layer on top of something more robust, like Jetty or Tomcat.

My thinking, for the most part, is that they probably did it better, so why should we try to compete with them. Where we can excel is as a simplified layer.

mpeterson2 replied:
I'm sure they have done it better as well. This is probably a good idea at some point.

sleepdeprecation commented on src/main/java/httpserver/MethodWrapper.java:
```diff
@@ -19,500 +21,147 @@
-        pathBuilder.append(paths[i]);
-      }
-
+    private static final String LANG_PATH = "java.lang.";
```
This can be deleted.

sleepdeprecation commented on src/main/java/httpserver/MethodWrapper.java:
```diff
@@ -19,500 +21,147 @@
+     *
+     * Paths should come in <code>/relative/path/to/match</code>. To use a
+     * variable in a path, it should come in
+     * <code>/path/with/{VariableType}</code> form, where VariableType is a
```
`s/{VariableType}/{VariableName}/g`

sleepdeprecation commented on src/main/java/httpserver/MethodWrapper.java:
```diff
@@ -19,500 +21,147 @@
+                if (i != paths.length - 1) {
+                    while (++i < paths.length && paths[i].isEmpty());
+                    if (i != paths.length - 1) {
+                        throw new HTTPException("{*} must be the final section of your path!");
```
Confirm this exception is thrown with a test somewhere!

mpeterson2 commented on src/main/java/httpserver/HTTPRequest.java:
```diff
@@ -25,509 +25,502 @@
-          [request type] [path] [protocol]
+    public HTTPHandler determineHandler() throws HTTPException {
+        if (router == null) {
+            return new DeathHandler();
```
This isn't quite specific to this commit, but maybe allowing the users to specify the handler to use here would be good? We could keep this as default but let users set it if they want.

sleepdeprecation replied:
I think this only happens if you don't have a router (`if (router == null)`), which would mean they haven't specified any handlers (or they did, but they didn't attach the router to the server).

If this ever happens we might want to throw some sort of `RuntimeException` or something, instead of saying everything's a-ok.

mpeterson2 commented on src/main/java/httpserver/MethodWrapper.java:
```diff
@@ -19,500 +21,147 @@
+     * @param path          The path that caused the method to be called. This is
+     *                      where variables come from.
+     */
+    public void invoke(HTTPRequest request, HTTPResponse response, List<String> calledPath) {
```
Maybe this could be split into a few methods. Maybe `getParams()` and `invoke()` where `invoke` just calls `getParams` instead of being so long.

sleepdeprecation replied:
I'm not sure I see what you're saying.

Do you think the `HTTPHandler#handle` method should do something like:

``` java
Map<String, String> params = method.extractParameters(calledPath);
request.mergeParams(params);

List<String> varargs = method.extractVarargs(calledPath);
request.mergeVarargs(varargs);

method.invoke(); // or maybe rename to method.handle() / method.call() ?
```

?

mpeterson2 commented:
I looked through it kinda quickly, but I didn't find anything that stood out as being terrible... so :+1:

sleepdeprecation committed:
Forgot to make sure MethodWrapper actually saves passed in Route.

This is a great time for #donisdumb...

This also adds some more tests, which actually caught the fact that the
MethodWrapper never saved the passed in Route. This is why tests are
important kids!

sleepdeprecation committed:
moved HTTP files to tmp. Done because OS X has a case-insensitive filesystem (wtf?)

sleepdeprecation committed:
And move back to where they should be, but with new names.

sleepdeprecation committed:
Rename all HTTP* classes to Http*

sleepdeprecation committed:
Fix janky tabbing.

sleepdeprecation committed:
Change HandlerTest to use new server as default handler setup.

sleepdeprecation commented on src/main/java/httpserver/HttpHandler.java:
```diff
@@ -0,0 +1,208 @@
+ * @see MessageHandler
+ */
+public abstract class HttpHandler {
+	private final HashMap<String, ArrayList<MethodWrapper>> routes = new HashMap<>();
```
This should be tabbed with spaces, not tabs.

sleepdeprecation commented on src/main/java/httpserver/HttpHandler.java:
```diff
@@ -0,0 +1,208 @@
+	 *
+	 * Additionally, primitives are not permited, because they're not classes in
+	 * the java.lang library. The three most common parameter types are String,
+	 * Integer, and Double.
```
This is super out of date, and the meat of the documentation should be moved to `#addRoute`.

sleepdeprecation commented on src/main/java/httpserver/HttpRequest.java:
```diff
@@ -0,0 +1,532 @@
+            setRequestType(DELETE_REQUEST_TYPE);
+        }
+        else {
+            throw new HttpException("Unexpected request type: " + splitty[0]);
```
1. Confirm that `GET`, `POST`, `HEAD`, and `DELETE` are the only "real" supported verbs.
2. Disregard that because it doesn't matter, and `HttpHandler` now supports _any_ request verb, official or not.

dkuntz2 commented on src/main/java/httpserver/HttpResponse.java:
```diff
@@ -0,0 +1,365 @@
+    public long getSize() {
+        return size;
+    }
+    public void setSize(long size) {
```
This should probably have some sort of check to make sure you don't say the content size is something like -400, because it can't be less than 0.

Alternatively, change line 149 to `if (getSize() >= 0)`.

sleepdeprecation commented on src/main/java/httpserver/HttpRouter.java:
```diff
@@ -0,0 +1,82 @@
+ * @see HttpHandler
+ * @see HttpRequest
+ */
+public class HttpRouter {
```
Thought: remove routers entirely, and switch to using Handlers for everything (and rename handlers to routers).

(Every instance of "Router" onward means what is currently called `HttpHandler`)

New routers would have path routing stuff built in (as they currently do), and have the added ability of adding a "sub router", which would work the same way `addHandler` works in this class.

Basically you'd have code that looks something like this:

``` java
HttpServer s = new HttpServer();
s.get("/", new Route() { 
    @Override public void handle(HttpRequest req, HttpResponse resp) { 
        resp.write("hi."); 
    }
});

HttpRouter subRouter = new HttpRouter();
subRouter.get("/", new Route() {
    @Override public void handle(HttpRequest req, HttpResponse resp) {
        resp.setBody("sub router!");
    }
});

s.addRouter("/sub", subRouter);

s.run();
```

The only two paths that would match would be:

```
"/"     => "hi."
"/sub/" => "sub router!"
```

Internally, the `Router#handle` method would first search to see if there's a subrouter for the first path segment, and try to use that. If there isn't, or if the remaining path segment doesn't match any of that router's routes, it'll go to the main routing mechanism currently in place. 

Hopefully that makes sense, it looks a little convoluted to me, and I'm the one trying to explain it.

sleepdeprecation replied:
This is a thought I had mainly because this class is now basically useless since for the most part future users should just be using `HttpServer` to do basically everything (see [`HandlerTest#serverSetup`](https://github.com/storytellersoftware/java-httpserver/blob/4539f034d529f159b08056da4b412d0746112d91/src/test/java/tests/HandlerTest.java#L26-L68) for an example).

sleepdeprecation commented on src/main/java/httpserver/HttpServer.java:
```diff
@@ -0,0 +1,210 @@
+ * and might not be the best base server for one to use. It exists solely to
+ * provide an existing mechanism for using the rest of the httpserver package.
+ */
+public class HttpServer extends HttpHandler {
```
This should probably become a layer on top of something else, like Jetty or Tomcat.

For reference, Spark uses Jetty as its backend.

What I'm thinking specifically is that the current API should remain intact, but the internals should use something more robust and less written by us.

sleepdeprecation commented on src/main/java/httpserver/HttpServer.java:
```diff
@@ -0,0 +1,210 @@
+     * Unless you specify the port with {@link HttpServer#setSocket()},
+     * the server will run on http://127.0.0.1:{@value #defaultPort}.
+     */
+    public void run() {
```
Why did we make this non-threaded?

sleepdeprecation commented on src/main/java/httpserver/MessageHandler.java:
```diff
@@ -6,9 +6,9 @@
+ * new HttpHandler.
  */
-public class MessageHandler extends HTTPHandler {
+public class MessageHandler extends HttpHandler {
```
Should we delete this / make it an example?

sleepdeprecation commented on src/main/java/httpserver/MethodWrapper.java:
```diff
@@ -1,14 +1,16 @@
+//      They're based on old, reflective, less conventional ways of doing
+//      things.
+//      Also, just clean this file up in general.
+//      And maybe rename it to "RouteWrapper?"
```
Is there any reason _not_ to just merge this with `Route`?

As in, make `Route` abstract (with the only abstract thing being a `handle` method), and have it basically do all of this?

So instead of writing

``` java
server.get("/", new Route() { /* override handle */ });
```

You'd write

``` java
server.get(new Route("/") { /* override handle */ });
```

sleepdeprecation commented on src/test/java/tests/HTTPRequestTest.java:
```diff
@@ -1,125 +1,120 @@
-  }
-  
-}
+//package tests;
```
Either delete or fix this.

sleepdeprecation commented on src/test/java/tests/HandlerTest.java:
```diff
@@ -1,70 +1,145 @@
+            }
+        });
+
 
```
Excess spaces.

sleepdeprecation commented on src/test/java/tests/ServerAsAppTest.java:
```diff
@@ -0,0 +1,39 @@
+import httpserver.HttpResponse;
+import httpserver.Route;
+
+public class ServerAsAppTest {
```
Delete this. It was basically merged in with `HandlerTest`.

sleepdeprecation commented on src/test/java/tests/mocks/MockClient.java:
```diff
@@ -12,65 +12,65 @@
 public class MockClient {
-  
+
   public static int DESIRED_PORT = 4444;
```
Is there some way to make a socket that doesn't use IP?

Basically, instead of having HttpRequest, HttpResponse, and MockClient use an actual ServerSocket, is there something that's basically just a text buffer that can be used instead?

sleepdeprecation commented on src/test/java/tests/HandlerTest.java:
```diff
@@ -1,70 +1,145 @@
+    private static HttpServer server;
+
+    @BeforeClass
+    public static void setupServer() {
```
Maybe we should think of some better way to abstract this out?

This is actually probably fine, I'm just thinking too far out there. We can figure out an abstraction layer later if we actually need it, and just write a little boilerplate for each test class.

sleepdeprecation committed:
Boyscouting.

sleepdeprecation committed:
Merge MethodWrapper and Route.

The main reason for doing this is that we needed a separate class to do
all of the magic reflection we previously used to make things work (or
add several hundred lines of code to the HttpHandler, which didn't seem
like a good idea), but because we no longer need to keep track of
separate methods, and have our routes just override the abstract
`handle` method, and pass around real objects, we're able to merge the
layer on top of the handle method with the object holding the handle
method.

Basically, this makes for cleaner, more conventional, and less magical
code. Less magic is always a good thing.
